### DISCUSSION TITLE
Why doesn’t Clangd trust project configurations?

### DISCUSSION THREAD

**Discussion by unimun** [clangd/clangd#2239] [Q&A]
It seems that this goes against the main benefit of using an external index server. The primary purpose of a remote index server is to reduce the resources and effort required by individual developers to index the repository. If the external index server were integrated with the project configuration, it would streamline onboarding for new developers, eliminating the need to manually set up user configurations.

An alternative approach that other tools like Git and VSCode adopt is to "trust the source directory." This approach allows the user to trust a repository as a whole without needing to explicitly specify details like the address of the remote index server. This way, all project-related information is centralized within project configurations, rather than being scattered across individual user configurations.

Looking at the code, there seems to be a bit of a tangle regarding the order of configuration compilation. Currently, Clangd prioritizes user configuration, meaning it is processed last and overrides project configurations. This makes it difficult to determine which project configuration is actually “trusted,” especially when overridden by user settings.

Perhaps we could revisit the order of precedence, as discussed in [this thread](https://github.com/clangd/clangd/discussions/1193). While the reasoning in the current implementation focuses on avoiding the accidental inclusion of project configurations in repositories, this also restricts the intentional addition of project-specific configurations, which could otherwise simplify setup for developers.

**Comment by HighCommander4** [comment]
Providing some additional context that wasn't obvious to me from the issue description: it looks like clangd currently [disallows](https://searchfox.org/llvm/rev/04cc492ec3e2792e3fa790750ff678baa90bd436/clang-tools-extra/clangd/ConfigCompile.cpp#345-351) specifying the [`Index.External`](https://clangd.llvm.org/config.html#external) config key (and only that one) in project config files.

The restriction was added in https://reviews.llvm.org/D100542; cc @kadircet and @sam-mccall for any thoughts on rationale / whether the restriction could be relaxed.

Regarding the order of precedence between user and project config, that seems like an orthogonal issue to me; is there a connection between the two that I'm missing?

**Reply by unimun** [reply]
You’re right that I may have mixed the trust issue with the precedence issue, which might have caused some confusion. My point is that to allow configuring a trusted remote index at the project config level, the precedence order would need to change to inner -> outer -> user, and the compile order should start with the user config.

This way, the user config would be read and reflected in the compiled configuration first, and the project config, which is processed later, would naturally override the compiled config, giving it higher precedence. This structure ensures that the project config can be validated as part of a trusted path and appropriately take precedence when necessary.

So, while trust and precedence are distinct issues, they are interconnected here because the compile order plays a critical role in determining whether project configs can reliably be trusted.

**Reply by i-ky** [reply]
> Regarding the order of precedence between user and project config, that seems like an orthogonal issue to me; is there a connection between the two that I'm missing?

Trust and order are indeed orthogonal, but `clangd` itself mixes these two concepts together. Check the error message at the [link](https://searchfox.org/llvm/rev/04cc492ec3e2792e3fa790750ff678baa90bd436/clang-tools-extra/clangd/ConfigCompile.cpp#345-351) you provided.

I've been puzzled by this error message myself, this resulted in #1329.

https://github.com/clangd/clangd/issues/908#issuecomment-952012788 and https://github.com/clangd/vscode-clangd/issues/318#issuecomment-1105034346 can shine some light on the rationale.

**Reply by unimun** [reply]
Thanks, @i-ky, for explaining the rationale behind the current configuration order!

I now understand that allowing project-level configurations to specify a remote index server introduces potential risks, such as the inclusion of a malicious server. However, I believe these concerns can be mitigated with an additional layer of security, separating the security considerations from the configuration structure.

For instance, `clangd` could maintain a file similar to `.ssh/known_hosts`, where remote index servers are registered along with their fingerprints. If a server changes, clangd could prompt the user to confirm whether to trust and add the new server. Additionally, the remote index server could support client authentication, which would not only prevent misuse but also help reduce server overload, especially in cases where its endpoint is unintentionally exposed in the remote repository.

**Reply by kadircet** [reply]
in terms of precedence, i'd still argue that user config should take over any configuration coming from the projects. These are "checked-in" configurations, that would reflect general style guides for a repository, not for the user. If the user wishes to see things differently, having the ability to override them via user config is crucial (instead of updating project configuration and pushing those changes to everyone working in a repo).

as for trust, sorry for lacking context in the review, but the scenario we're trying to prevent is some remote keylogger :)
a possible attack vector is, someone might set up a "malicious" .clangd config in a repository, and any user that starts working on that project will make code-completion requests at every keystroke to a remote server outside of their control, implicitly.

we had some plans to extend definition of "trusted config" by asking for users to bless `(config-key, value)` pairs, on capable editors (e.g. ones that can do prompts) and persist that consent, but never got to it. as a result only trusted config ATM is user config, which has to be explicitly managed by the user (or a malicious actor outside of clangd's influence, hopefully).

so i'd still advise against extending trust to project configurations, until we have a way to get user blessing.
